File: pkg/ot/ot.go
package ot

import "fmt"

type Op interface{ opKind() }

type RetainOp struct{ N int }
type InsertOp struct{ Text string }
type DeleteOp struct{ N int }

func (RetainOp) opKind() {}
func (InsertOp) opKind() {}
func (DeleteOp) opKind() {}

type Operation struct {
	Ops          []Op
	BaseLength   int
	TargetLength int
	ClientID     int
}

func NewOperation() *Operation {
	return &Operation{}
}

func (o *Operation) Retain(n int) *Operation {
	if n == 0 {
		return o
	}
	o.BaseLength += n
	o.TargetLength += n
	if len(o.Ops) > 0 {
		if last, ok := o.Ops[len(o.Ops)-1].(RetainOp); ok {
			o.Ops[len(o.Ops)-1] = RetainOp{N: last.N + n}
			return o
		}
	}
	o.Ops = append(o.Ops, RetainOp{N: n})
	return o
}

func (o *Operation) Insert(text string) *Operation {
	if text == "" {
		return o
	}
	o.TargetLength += len(text)
	last := len(o.Ops) - 1
	if last >= 0 {
		if ins, ok := o.Ops[last].(InsertOp); ok {
			o.Ops[last] = InsertOp{Text: ins.Text + text}
			return o
		}
		if _, ok := o.Ops[last].(DeleteOp); ok {
			if last-1 >= 0 {
				if ins, ok := o.Ops[last-1].(InsertOp); ok {
					o.Ops[last-1] = InsertOp{Text: ins.Text + text}
					return o
				}
			}
			// insert before the delete
			o.Ops = append(o.Ops, nil)
			o.Ops[last+1] = o.Ops[last]
			o.Ops[last] = InsertOp{Text: text}
			return o
		}
	}
	o.Ops = append(o.Ops, InsertOp{Text: text})
	return o
}

func (o *Operation) Delete(n int) *Operation {
	if n == 0 {
		return o
	}
	o.BaseLength += n
	if len(o.Ops) > 0 {
		if last, ok := o.Ops[len(o.Ops)-1].(DeleteOp); ok {
			o.Ops[len(o.Ops)-1] = DeleteOp{N: last.N + n}
			return o
		}
	}
	o.Ops = append(o.Ops, DeleteOp{N: n})
	return o
}

func (o *Operation) Apply(str string) (string, error) {
	if len(str) != o.BaseLength {
		return "", fmt.Errorf("the operation's base length must be equal to the string's length")
	}
	result := make([]byte, 0, o.TargetLength)
	strIndex := 0
	for _, op := range o.Ops {
		switch v := op.(type) {
		case RetainOp:
			if strIndex+v.N > len(str) {
				return "", fmt.Errorf("operation can't retain more characters than are left in the string")
			}
			result = append(result, str[strIndex:strIndex+v.N]...)
			strIndex += v.N
		case InsertOp:
			result = append(result, v.Text...)
		case DeleteOp:
			strIndex += v.N
		}
	}
	if strIndex != len(str) {
		return "", fmt.Errorf("the operation didn't operate on the whole string")
	}
	return string(result), nil
}

func Transform(op1, op2 *Operation) (*Operation, *Operation, error) {
	if op1.BaseLength != op2.BaseLength {
		return nil, nil, fmt.Errorf("both operations have to have the same base length")
	}

	op1Prime := &Operation{ClientID: op1.ClientID}
	op2Prime := &Operation{ClientID: op2.ClientID}

	ops1 := op1.Ops
	ops2 := op2.Ops
	i1, i2 := 0, 0

	var o1, o2 Op
	nextO1 := func() {
		if i1 < len(ops1) {
			o1 = ops1[i1]
			i1++
		} else {
			o1 = nil
		}
	}
	nextO2 := func() {
		if i2 < len(ops2) {
			o2 = ops2[i2]
			i2++
		} else {
			o2 = nil
		}
	}

	nextO1()
	nextO2()

	for {
		if o1 == nil && o2 == nil {
			break
		}

		_, o1IsInsert := o1.(InsertOp)
		_, o2IsInsert := o2.(InsertOp)

		if o1IsInsert && (!o2IsInsert || op1.ClientID > op2.ClientID) {
			op1Prime.Insert(o1.(InsertOp).Text)
			op2Prime.Retain(len(o1.(InsertOp).Text))
			nextO1()
			continue
		}
		if o2IsInsert {
			op1Prime.Retain(len(o2.(InsertOp).Text))
			op2Prime.Insert(o2.(InsertOp).Text)
			nextO2()
			continue
		}

		if o1 == nil {
			return nil, nil, fmt.Errorf("cannot compose operations: first operation is too short")
		}
		if o2 == nil {
			return nil, nil, fmt.Errorf("cannot compose operations: first operation is too long")
		}

		switch v1 := o1.(type) {
		case RetainOp:
			switch v2 := o2.(type) {
			case RetainOp:
				minL := min(v1.N, v2.N)
				v1.N -= minL
				v2.N -= minL
				if v1.N == 0 {
					nextO1()
				} else {
					o1 = v1
				}
				if v2.N == 0 {
					nextO2()
				} else {
					o2 = v2
				}
				op1Prime.Retain(minL)
				op2Prime.Retain(minL)
			case DeleteOp:
				minL := min(v1.N, v2.N)
				v1.N -= minL
				v2.N -= minL
				if v1.N == 0 {
					nextO1()
				} else {
					o1 = v1
				}
				if v2.N == 0 {
					nextO2()
				} else {
					o2 = v2
				}
				op2Prime.Delete(minL)
			}
		case DeleteOp:
			switch v2 := o2.(type) {
			case DeleteOp:
				minL := min(v1.N, v2.N)
				v1.N -= minL
				v2.N -= minL
				if v1.N == 0 {
					nextO1()
				} else {
					o1 = v1
				}
				if v2.N == 0 {
					nextO2()
				} else {
					o2 = v2
				}
			case RetainOp:
				minL := min(v1.N, v2.N)
				v1.N -= minL
				v2.N -= minL
				if v1.N == 0 {
					nextO1()
				} else {
					o1 = v1
				}
				if v2.N == 0 {
					nextO2()
				} else {
					o2 = v2
				}
				op1Prime.Delete(minL)
			}
		default:
			return nil, nil, fmt.Errorf("the two operations aren't compatible")
		}
	}

	return op1Prime, op2Prime, nil
}

// wire format: positive int = retain, negative int = delete, string = insert

type OperationData struct {
	ClientID     int   `json:"cid"`
	Ops          []any `json:"ops"`
	BaseLength   int   `json:"base"`
	TargetLength int   `json:"target"`
}

func SerializeOperation(op *Operation) OperationData {
	ops := make([]any, len(op.Ops))
	for i, o := range op.Ops {
		switch v := o.(type) {
		case RetainOp:
			ops[i] = v.N
		case InsertOp:
			ops[i] = v.Text
		case DeleteOp:
			ops[i] = -v.N
		}
	}
	return OperationData{
		ClientID:     op.ClientID,
		Ops:          ops,
		BaseLength:   op.BaseLength,
		TargetLength: op.TargetLength,
	}
}

func DeserializeOperation(data OperationData) (*Operation, error) {
	op := &Operation{
		ClientID:     data.ClientID,
		BaseLength:   data.BaseLength,
		TargetLength: data.TargetLength,
	}
	for _, raw := range data.Ops {
		switch v := raw.(type) {
		case float64:
			if v > 0 {
				op.Ops = append(op.Ops, RetainOp{N: int(v)})
			} else if v < 0 {
				op.Ops = append(op.Ops, DeleteOp{N: int(-v)})
			}
		case string:
			op.Ops = append(op.Ops, InsertOp{Text: v})
		default:
			return nil, fmt.Errorf("unexpected op type in deserialization: %T", raw)
		}
	}
	return op, nil
}

File: assets/src/ot.ts
export type RetainOp = { type: "retain"; n: number };
export type InsertOp = { type: "insert"; text: string };
export type DeleteOp = { type: "delete"; n: number };
export type Op = RetainOp | InsertOp | DeleteOp;

export function retain(n: number): RetainOp {
  return { type: "retain", n };
}
export function insert(text: string): InsertOp {
  return { type: "insert", text };
}
export function del(n: number): DeleteOp {
  return { type: "delete", n };
}

export function isRetain(op: Op): op is RetainOp {
  return op.type === "retain";
}
export function isInsert(op: Op): op is InsertOp {
  return op.type === "insert";
}
export function isDelete(op: Op): op is DeleteOp {
  return op.type === "delete";
}

export class Operation {
  ops: Op[] = [];
  baseLength = 0;
  targetLength = 0;
  clientID: number = 0;

  retain(n: number): this {
    if (n === 0) return this;
    this.baseLength += n;
    this.targetLength += n;
    const last = this.ops[this.ops.length - 1];
    if (last && isRetain(last)) {
      last.n += n;
    } else {
      this.ops.push(retain(n));
    }
    return this;
  }

  insert(text: string): this {
    if (text === "") return this;
    this.targetLength += text.length;
    const last = this.ops[this.ops.length - 1];
    const secondLast = this.ops[this.ops.length - 2];
    if (last && isInsert(last)) {
      last.text += text;
    } else if (last && isDelete(last)) {
      if (secondLast && isInsert(secondLast)) {
        secondLast.text += text;
      } else {
        this.ops.splice(this.ops.length - 1, 0, insert(text));
      }
    } else {
      this.ops.push(insert(text));
    }
    return this;
  }

  delete(n: number): this {
    if (n === 0) return this;
    this.baseLength += n;
    const last = this.ops[this.ops.length - 1];
    if (last && isDelete(last)) {
      last.n += n;
    } else {
      this.ops.push(del(n));
    }
    return this;
  }

  apply(str: string): string {
    if (str.length !== this.baseLength) {
      throw new Error(
        "The operation's base length must be equal to the string's length",
      );
    }
    const parts: string[] = [];
    let strIndex = 0;
    for (const op of this.ops) {
      if (isRetain(op)) {
        if (strIndex + op.n > str.length) {
          throw new Error(
            "Operation can't retain more characters than are left in the string.",
          );
        }
        parts.push(str.slice(strIndex, strIndex + op.n));
        strIndex += op.n;
      } else if (isInsert(op)) {
        parts.push(op.text);
      } else {
        strIndex += op.n;
      }
    }
    if (strIndex !== str.length) {
      throw new Error("The operation didn't operate on the whole string.");
    }
    return parts.join("");
  }
}

export function transform(
  operation1: Operation,
  operation2: Operation,
): [Operation, Operation] {
  if (operation1.baseLength !== operation2.baseLength) {
    throw new Error("Both operations have to have the same base length");
  }

  const op1Prime = new Operation();
  op1Prime.clientID = operation1.clientID;
  const op2Prime = new Operation();
  op2Prime.clientID = operation2.clientID;

  const ops1 = operation1.ops;
  const ops2 = operation2.ops;
  let i1 = 0;
  let i2 = 0;
  let o1 = ops1[i1++];
  let o2 = ops2[i2++];

  while (true) {
    if (!o1 && !o2) break;

    if (
      o1 &&
      isInsert(o1) &&
      (!o2 || !isInsert(o2) || operation1.clientID > operation2.clientID)
    ) {
      op1Prime.insert(o1.text);
      op2Prime.retain(o1.text.length);
      o1 = ops1[i1++];
      continue;
    }
    if (o2 && isInsert(o2)) {
      op1Prime.retain(o2.text.length);
      op2Prime.insert(o2.text);
      o2 = ops2[i2++];
      continue;
    }

    if (!o1)
      throw new Error(
        "Cannot compose operations: first operation is too short.",
      );
    if (!o2)
      throw new Error(
        "Cannot compose operations: first operation is too long.",
      );

    if (isRetain(o1) && isRetain(o2)) {
      const minL = Math.min(o1.n, o2.n);
      o1.n -= minL;
      o2.n -= minL;
      if (o1.n === 0) o1 = ops1[i1++];
      if (o2.n === 0) o2 = ops2[i2++];
      op1Prime.retain(minL);
      op2Prime.retain(minL);
    } else if (isDelete(o1) && isDelete(o2)) {
      const minL = Math.min(o1.n, o2.n);
      o1.n -= minL;
      o2.n -= minL;
      if (o1.n === 0) o1 = ops1[i1++];
      if (o2.n === 0) o2 = ops2[i2++];
    } else if (isDelete(o1) && isRetain(o2)) {
      const minL = Math.min(o1.n, o2.n);
      o1.n -= minL;
      o2.n -= minL;
      if (o1.n === 0) o1 = ops1[i1++];
      if (o2.n === 0) o2 = ops2[i2++];
      op1Prime.delete(minL);
    } else if (isRetain(o1) && isDelete(o2)) {
      const minL = Math.min(o1.n, o2.n);
      o1.n -= minL;
      o2.n -= minL;
      if (o1.n === 0) o1 = ops1[i1++];
      if (o2.n === 0) o2 = ops2[i2++];
      op2Prime.delete(minL);
    } else {
      throw new Error("The two operations aren't compatible");
    }
  }

  return [op1Prime, op2Prime];
}

// Wire format: positive int = retain, negative int = delete, string = insert

export type WireOp = string | number;

export interface OperationData {
  cid: number;
  ops: WireOp[];
  base: number;
  target: number;
}

export function serializeOperation(op: Operation): OperationData {
  return {
    cid: op.clientID,
    base: op.baseLength,
    target: op.targetLength,
    ops: op.ops.map((o) => {
      if (isRetain(o)) return o.n;
      if (isInsert(o)) return o.text;
      return -o.n;
    }),
  };
}

export function deserializeOperation(data: OperationData): Operation {
  const op = new Operation();
  op.clientID = data.cid;
  op.baseLength = data.base;
  op.targetLength = data.target;
  for (const raw of data.ops) {
    if (typeof raw === "string") {
      op.ops.push(insert(raw));
    } else if (raw > 0) {
      op.ops.push(retain(raw));
    } else if (raw < 0) {
      op.ops.push(del(-raw));
    } else {
      throw new Error("Invalid op in deserialization: 0 is not a valid op");
    }
  }
  return op;
}
